File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
  public int val;
  public Node left;
  public Node right;
  public Node next;

  public Node() {}
  
  public Node(int _val) {
    val = _val;
  }

  public Node(int _val, Node _left, Node _right, Node _next) {
    val = _val;
    left = _left;
    right = _right;
    next = _next;
  }
};
*/

class AuxNode {
  private int level;
  private Node node;
  
  public AuxNode(int level, Node node) {
    this.level = level;
    this.node = node;
  }
  
  public int getLevel() {
    return level;
  }
  
  public Node getNode() {
    return node;
  }
}

class Solution {
  public Node connect(Node root) {
    List<AuxNode> queue = new ArrayList<AuxNode>();
    if (root != null) queue.add(new AuxNode(0, root));
    List<List<Node>> levels = new ArrayList();
    while (!queue.isEmpty()) {
      AuxNode curr = queue.remove(0);
      int level = curr.getLevel();
      Node node = curr.getNode();
      if (level > levels.size() - 1) levels.add(new ArrayList<Node>());
      levels.get(level).add(node);
      if (node.left != null) queue.add(new AuxNode(level + 1, node.left));
      if (node.right != null) queue.add(new AuxNode(level + 1, node.right));
    }
    Node currNode;
    while (!levels.isEmpty()) {
      List<Node> currLevel = levels.remove(0);
      currNode = null;
      Node last;
      while (!currLevel.isEmpty()) {
        last = currLevel.remove(currLevel.size() - 1);
        last.next = currNode;
        currNode = last;
      }
    }
    return root;
  }
}

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.go
/**
 * Definition for a Node.
 * type Node struct {
 *   Val int
 *   Left *Node
 *   Right *Node
 *   Next *Node
 * }
 */
type AuxNode struct {
  Level int
  Node *Node
}

func connect(root *Node) *Node {
  queue := []*AuxNode{}
  if root != nil {
    queue = append(queue, &AuxNode{
      Level: 0,
      Node: root,
    })
  }
  levels := [][]*Node{}
  for len(queue) > 0 {
    curr := queue[0]
    queue = queue[1:]
    level, node := (*curr).Level, (*curr).Node
    if level > len(levels) - 1 {
      levels = append(levels, []*Node{})
    }
    levels[level] = append(levels[level], node)
    if (*node).Left != nil {
      queue = append(queue, &AuxNode{
        Level: level + 1,
        Node: (*node).Left,
      })
    }
    if (*node).Right != nil {
      queue = append(queue, &AuxNode{
        Level: level + 1,
        Node: (*node).Right,
      })
    }
  }
  for len(levels) > 0 {
    currLevel := levels[0]
    var currNode *Node
    levels = levels[1:]
    var last *Node
    for len(currLevel) > 0 {
      last = currLevel[len(currLevel) - 1]
      currLevel = currLevel[:len(currLevel) - 1]
      (*last).Next = currNode
      currNode = last
    }
  }
  return root
}